### ISSUE TITLE
status symbols for `flycheck-report-buffer-checker-status`

### ISSUE THREAD

**Issue by gdkrmr** [flycheck/flycheck#1295]
The `callback` of a linter defined by `flycheck-define-generic-checker` currently has 4 possible status symbols, I copied them from the docstring for convenience:

> ‘errored’
>      The syntax checker has errored.  DATA is an optional error
>      message.
>      This report finishes the current syntax check.
> 
> ‘interrupted’
>      The syntax checker was interrupted.  DATA is ignored.
>      This report finishes the current syntax check.
> 
> ‘finished’
>      The syntax checker has finished with a proper error report
>      for the current buffer.  DATA is the (potentially empty)
>      list of ‘flycheck-error’ objects reported by the syntax
>      check.
>      This report finishes the current syntax check.
> 
> ‘suspicious’
>      The syntax checker encountered a suspicious state, which the
>      user needs to be informed about.  DATA is an optional
>      message.

Please correct me if I am wrong:
If I start an asynchronous process and want to retrieve the errors in a `sentinel`, then the only symbol I can use for my callback in the main checker function (the one in `:start`) is `'suspicious`, else the syntax check will terminate and all the other calls to `callback` will be ignored. 
This is not really intuitive, I think that there should be an additional symbol for this, e.g. `'started`, especially because the documentation for `flycheck-define-generic-checker` mentions exactly this case.

> FUNCTION may be synchronous or asynchronous, i.e. it may
> call the given callback either immediately, or at some later
> point (e.g. from a process sentinel).

Here is the model I used, I hope that makes everything a little bit clearer:

```elisp
(defun query-server (checker callback)
  ...
  ;; spawn the process that makes the actual query
  (make-process
    ...
    :sentinel (lambda (process event)
                   ...
                   (funcall callback 'finished errors))
    ...
    )
  ...

  (funcall callback 'suspicious "check started")
```

**Comment by cpitclaudel** [comment]
I think the thing you're missing is that you don't need to call the callback in the main function :)  That is, just don't call `callback` at all until you're in the sentinel.

**Comment by gdkrmr** [comment]
the documentation says that I need to run the callback at least once else `flycheck` will be in an undefined state, so in case the sentinel never gets called I will crash `flycheck`.

**Comment by cpitclaudel** [comment]
> the documentation says that I need to run the callback at least once else flycheck will be in an undefined state

No, not really. There's no mention of undefined state AFAICT. The doc says this:

```
     A syntax checker _must_ call CALLBACK at least once with a
     STATUS that finishes the current syntax checker.  Otherwise
     Flycheck gets stuck at the current syntax check with this
     syntax checker.
```

This doesn't say that you need to call the callback immediately and then a second time later.  It just says that until you complete the current check, a new one won't be started.

> so in case the sentinel never gets called I will crash flycheck.

You won't crash it; you'll just prevent the next check from starting.  Which makes sense, since you won't have told flycheck about the previous one completing.

So indeed, you need to ensure that your sentinel gets called.  Is that an issue?

**Comment by gdkrmr** [comment]
I wait for the connection to get closed by the lint server, which causes the sentinel to get called. I am not sure what happens if the server gets stuck for some reason.

**Comment by cpitclaudel** [comment]
> I am not sure what happens if the server gets stuck for some reason.

In that case the syntax check gets stuck too, waiting for results.  I think that's not a bad thing, is it?

**Comment by gdkrmr** [comment]
ok, I'll simply remove that first callback then, thanks for the help!

**Comment by cpitclaudel** [comment]
Sure! Let us know if you run into issues :)
